src/Peachpie.Library/OpenSSL.cs

```diff
@@ -0,0 +1,741 @@
+            }
+
+            if (reminder != encoded.Length - 1)
+                builder.AppendLine(encoded.Substring(reminder, encoded.Length - reminder));
+
+            builder.Append("-----END CERTIFICATE-----");
+
+            return builder.ToString();
+        }
+
+        /// <summary>
+        /// Calculates the fingerprint, or digest, of a given X.509 certificate
+        /// </summary>
+        /// <param name="ctx">Context of the script.</param>
+        /// <param name="x509">Path to file with PEM encoded certificate or a string containing the content of a certificate or X509Resource</param>
+        /// <param name="hash_algorithm">The digest method or hash algorithm to use, e.g. "sha256", one of openssl_get_md_methods().</param>
+        /// <param name="raw_output">When set to TRUE, outputs raw binary data. FALSE outputs lowercase hexits.</param>
+        /// <returns>Returns a string containing the calculated certificate fingerprint as lowercase hexits unless raw_output is set to TRUE in which case the raw binary representation of the message digest is returned.</returns>
+        [return: CastToFalse]
+        public static PhpString openssl_x509_fingerprint(Context ctx, PhpValue x509, string hash_algorithm = "sha1", bool raw_output = false)
+        {
+            var resource = ParseX509Certificate(ctx, x509);
+            if (resource == null)
+                return null;
+
+            return openssl_digest(ctx, new PhpString(resource.Certificate.Export(X509ContentType.Cert)), hash_algorithm, raw_output);
+        }
+
+        /// <summary>
+        /// Stores x509 into a string named by output in a PEM encoded form
+        /// </summary>
+        /// <param name="ctx">Context of the script.</param>
+        /// <param name="x509">Path to file with PEM encoded certificate or a string containing the content of a certificate or X509Resource</param>
+        /// <param name="output">On success, this will hold the PEM.</param>
+        /// <param name="notext">The optional parameter notext affects the verbosity of the output; if it is FALSE, then additional human-readable information is included in the output. The default value of notext is TRUE.</param>
+        /// <returns>Returns TRUE on success or FALSE on failure.</returns>
+        public static bool openssl_x509_export(Context ctx, PhpValue x509, ref string output, bool notext = true)
+        {
+            var resource = ParseX509Certificate(ctx, x509);
+            if (resource == null)
+                return false;
+
+            output = Export(resource, notext);
+
+            return true;
+        }
+
+        /// <summary>
+        /// Frees the certificate associated with the specified x509cert resource from memory.
+        /// </summary>
+        public static void openssl_x509_free(PhpResource x509cert )
+        {
+            if (x509cert is X509Resource h && x509cert.IsValid)
+                h.FreeCertificate();
```

Comment:
```c#
h?.Dispose()
```

Reply:
Is it `?` necesassary ? I think, it will not be null because of if statement above

Reply:
`?` is not, calling `Dispose` instead of `FreeCertificate` is suggested :)